x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCancelJobModelSnapshotUpgradeAction.java

```diff
@@ -0,0 +1,146 @@
+import org.elasticsearch.xpack.core.ml.action.CancelJobModelSnapshotUpgradeAction.Response;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.job.snapshot.upgrade.SnapshotUpgradeTaskParams;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.job.persistence.JobConfigProvider;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class TransportCancelJobModelSnapshotUpgradeAction extends HandledTransportAction<Request, Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportCancelJobModelSnapshotUpgradeAction.class);
+
+    private final JobConfigProvider jobConfigProvider;
+    private final ClusterService clusterService;
+    private final PersistentTasksService persistentTasksService;
+
+    @Inject
+    public TransportCancelJobModelSnapshotUpgradeAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        JobConfigProvider jobConfigProvider,
+        ClusterService clusterService,
+        PersistentTasksService persistentTasksService
+    ) {
+        super(CancelJobModelSnapshotUpgradeAction.NAME, transportService, actionFilters, Request::new);
+        this.jobConfigProvider = jobConfigProvider;
+        this.clusterService = clusterService;
+        this.persistentTasksService = persistentTasksService;
+    }
+
+    @Override
+    public void doExecute(Task task, Request request, ActionListener<Response> listener) {
+
+        logger.debug("[{}] cancel model snapshot [{}] upgrades", request.getJobId(), request.getSnapshotId());
+
+        // 2. Now that we have the job IDs, find the relevant model snapshot upgrade tasks
+        ActionListener<List<Job.Builder>> expandIdsListener = ActionListener.wrap(jobs -> {
+            SimpleIdsMatcher matcher = new SimpleIdsMatcher(request.getSnapshotId());
+            Set<String> jobIds = jobs.stream().map(Job.Builder::getId).collect(Collectors.toSet());
+            PersistentTasksCustomMetadata tasksInProgress = clusterService.state().metadata().custom(PersistentTasksCustomMetadata.TYPE);
+            // allow_no_match plays no part here. The reason is that we have a principle that stopping
+            // a stopped entity is a no-op, and upgrades that have already completed won't have a task.
+            // This is a bit different to jobs and datafeeds, where the entity continues to exist even
+            // after it's stopped. Upgrades cease to exist after they're stopped so the match validation
+            // cannot be as thorough.
+            List<PersistentTasksCustomMetadata.PersistentTask<?>> upgradeTasksToCancel = MlTasks.snapshotUpgradeTasks(tasksInProgress)
+                .stream()
+                .filter(t -> jobIds.contains(((SnapshotUpgradeTaskParams) t.getParams()).getJobId()))
```

Comment:
`expandJobs` handles `allow_no_match`.  It's impossible in this action for `allow_no_match` to work with the snapshot IDs, because the upgrade tasks are ephemeral and stopping a stopped thing is not an error.  But `allow_no_match` _can_ work with the job IDs.

This also ties in with https://github.com/elastic/elasticsearch/pull/81831#discussion_r772515647 because if we were only ever going to use this action internally then it wouldn't be so important to support `allow_no_match` at all.  But I can imagine in the future we might want to call this action from an enhanced UI.  We'll need to add a REST endpoint at that time and it's nice if the machinery of `allow_no_match` is available.  The way `allow_no_match` is handled for jobs in this action is the same as in the recently added model snapshot upgrades stats action.

Reply:
There is a difference to the GET stats API which enforces that the snapshot IDs are also matched. 

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetJobModelSnapshotsUpgradeStatsAction.java#L110

Reply:
I think what I've done is as sensible as it can be given the way the persistent tasks appear and disappear.

Consider these scenarios:

### 1
- User asks for stats for job foo, upgrade of snapshot 1
- Job foo doesn't exist
- We return not found
- This isn't controversial

### 2
- User asks for stats for job foo, upgrade of snapshot 1
- Job foo exists but snapshot 1 isn't being upgraded
- We return not found
- This one is potentially controversial, as the snapshot upgrade might only just have finished, but since it's not running we cannot return any stats.  The only things we could do to make this better would be (1) persist stats for snapshot upgrades that have finished or (2) make upgrading a snapshot into a multi-step process where you first open the upgrade, then later close it (like you do for jobs).  Both of these things bring considerable extra baggage to the upgrade process.

### 3
- User asks for stats for job foo, upgrade of snapshot 1
- Job foo exists and snapshot 1 is being upgraded
- We return the stats
- This isn't controversial

### 4
- User asks to cancel upgrade of snapshot 1 for job foo
- Job foo doesn't exist
- We return not found
- This isn't controversial

### 5
- User asks to cancel upgrade of snapshot 1 for job foo
- Job foo exists but snapshot 1 isn't being upgraded
- We do nothing and return success
- This one is potentially controversial, but the end state is what the user wanted and the snapshot upgrade _might_ have just finished in between the user listing the upgrades in progress and asking to cancel this one.  This is also consistent with closing a closed job is not an error and stopping a stopped datafeed is not an error.

### 6
- User asks to cancel upgrade of snapshot 1 for job foo
- Job foo exists and snapshot 1 is being upgraded
- We cancel the upgrade and return success
- This isn't controversial

I don't think there's a better way.